_posts/2020-02-11-spring-boot-openapi.md

```diff
@@ -0,0 +1,394 @@
+                    <delegatePattern>true</delegatePattern>
+                </configOptions>
+            </configuration>
+        </execution>
+    </executions>
+</plugin>
+```
+
+You can see the full `pom.xml` file [on GitHub](https://github.com/thombergs/code-examples/tree/master/spring-boot/spring-boot-openapi/specification/pom.xml).
+
+For this tutorial, we'll use the `spring` generator. The [OpenAPI Maven plugin](https://github.com/OpenAPITools/openapi-generator/tree/master/modules/openapi-generator-maven-plugin) comes with a lot of options. To see the full set of options you can go over to the [official site](https://github.com/OpenAPITools/openapi-generator/tree/master/modules/openapi-generator-maven-plugin) of the plugin. 
+
+Simply running the command `mvnw install` will generate code that implements our OpenAPI specification!
+
+If you take a quick look under `target/generated-sources/openapi/src/main/java/io/reflectoring/model`, you will find the User model defined in `openapi.yml`!
+
+```java
+@javax.annotation.Generated(value = "org.openapitools.codegen.languages.SpringCodegen", date = "2020-02-07T19:50:29.578+01:00[Europe/Berlin]")
+
+public class User   {
+  @JsonProperty("id")
+  private Long id;
+
+  @JsonProperty("username")
+  private String username;
+
+  @JsonProperty("firstName")
+  private String firstName;
+
+  @JsonProperty("lastName")
+  private String lastName;
+
+  @JsonProperty("email")
+  private String email;
+
+  @JsonProperty("password")
+  private String password;
+
+  @JsonProperty("phone")
+  private String phone;
+
+  @JsonProperty("userStatus")
+  private Integer userStatus;
+}
+// omit getters and setters
+```
+
+Now let's implement the `UserApi`, which comes from our generated specification.
+
+```java
+@RestController
+public class UserController implements UserApi {
+}
+```
+
+If we test our API without having implemented a method of `UserApi`, we'll get a 501 response (Not Implemented). Why?
+
+```shell script
+curl -I http://localhost:8080/user/Petros
+HTTP/1.1 501
+Content-Length: 0
+```
+
+Because in UserApi says so:
+
+```java
+default ResponseEntity<User> getUserByName(String username) {
+    getRequest().ifPresent(request -> {
+        for (MediaType mediaType: MediaType.parseMediaTypes(request.getHeader("Accept"))) {
+            if (mediaType.isCompatibleWith(MediaType.valueOf("application/json"))) {
+                String exampleString = "{ \"firstName\" : \"firstName\", \"lastName\" : \"lastName\", \"password\" : \"password\", \"userStatus\" : 6, \"phone\" : \"phone\", \"id\" : 0, \"email\" : \"email\", \"username\" : \"username\" }";
+                ApiUtil.setExampleResponse(request, "application/json", exampleString);
+                break;
+            }
+        }
+    });
+    return new ResponseEntity<>(HttpStatus.NOT_IMPLEMENTED);
+
+}
+```
+
+Let's now just implement our actual GET method:
+
+```java
+@Override
+public ResponseEntity<User> getUserByName(String username) {
```

Comment:
Again, please include the class, otherwise the reader has to guess that this is in the `UserController` class.